include/arch/arm/aarch32/misc.h

```diff
@@ -31,10 +32,35 @@ static ALWAYS_INLINE void arch_nop(void)
 	__asm__ volatile("nop");
 }
 
+#if defined(CONFIG_USERSPACE)
+#if defined(CONFIG_CPU_CORTEX_M)
+#include <arch/arm/aarch32/cortex_m/cmsis.h>
+#elif defined(CONFIG_CPU_CORTEX_R)
+#include <arch/arm/aarch32/cortex_a_r/cmsis.h>
+#endif
+
+static inline bool z_arm_thread_is_in_user_mode(void)
+{
+	uint32_t value;
+
+#if defined(CONFIG_CPU_CORTEX_R)
+	/*
+	 * For Cortex-R, the mode (lower 5) bits will be 0x10 for user mode.
+	 */
+	value = __get_CPSR();
+	return ((value & CPSR_M_Msk) == CPSR_M_USR);
+#else
```

Comment:
Strictly speaking we need an #else if CORTEX_M here, and an #else with some error information